feature/bgp/addpath/otg_tests/route_propagation_test/route_propagation_test.go

```diff
@@ -138,6 +208,85 @@ func (d *dutData) AwaitBGPEstablished(t *testing.T, dut *ondatra.DUTDevice) {
 	t.Log("BGP sessions established")
 }
 
+func verifyOTGBGPTelemetry(t *testing.T, otg *otg.OTG, c gosnappi.Config, state string) {
+	for _, d := range c.Devices().Items() {
+		for _, ip := range d.Bgp().Ipv4Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := otg.Telemetry().BgpPeer(configPeer.Name())
+				_, ok := nbrPath.SessionState().Watch(t, time.Minute,
+					func(val *otgtelemetry.QualifiedE_BgpPeer_SessionState) bool {
+						return val.IsPresent() && val.Val(t).String() == state
+					}).Await(t)
+				if !ok {
+					fptest.LogYgot(t, "BGP reported state", nbrPath, nbrPath.Get(t))
+					t.Errorf("No BGP neighbor formed for peer %s", configPeer.Name())
+				}
+			}
+		}
+		for _, ip := range d.Bgp().Ipv6Interfaces().Items() {
+			for _, configPeer := range ip.Peers().Items() {
+				nbrPath := otg.Telemetry().BgpPeer(configPeer.Name())
+				_, ok := nbrPath.SessionState().Watch(t, time.Minute,
+					func(val *otgtelemetry.QualifiedE_BgpPeer_SessionState) bool {
+						return val.IsPresent() && val.Val(t).String() == state
+					}).Await(t)
+				if !ok {
+					fptest.LogYgot(t, "BGP reported state", nbrPath, nbrPath.Get(t))
+					t.Errorf("No BGP neighbor formed for peer %s", configPeer.Name())
+				}
+			}
+		}
+	}
+}
+
+type OTGBGPPrefix struct {
+	PeerName     string
+	Address      string
+	PrefixLength uint32
+}
+
+func checkOTGBGP4Prefix(t *testing.T, otg *otg.OTG, config gosnappi.Config, expectedOTGBGPPrefix OTGBGPPrefix) bool {
+	bgpPrefixes := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Get(t)
```

Comment:
The intent was to use the Watch/Await combo like in the feature/bgp/policybase test [here](https://github.com/open-traffic-generator/featureprofiles/blob/62dd6ddec33819b39faf429ce20fa666427a3f96/feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go#L470) but there is no **UnicastIpv4Prefix** function [here](https://github.com/open-traffic-generator/featureprofiles/blob/62dd6ddec33819b39faf429ce20fa666427a3f96/feature/bgp/addpath/otg_tests/route_propagation_test/route_propagation_test.go#L253) to support the Watch/Await combo

Reply:
I'm not that familiar with the watch API, but it sounds like the Any query does not support the Watch API?

Reply:
Yes, with "any" we don't have something to watch for and I know there were some problems with having "non-any" functions there like **UnicastIpv4Prefix**. I will ask someone to provide more details here.

Reply:
it is possible to Watch an Any path. In this case, the predicate is called on every Ipv4Prefix received from the OTG.
```
	otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch(t, time.Minute, func(bgpPrefix *otgstructs.QualifiedBgpPeer_UnicastIpv4Prefix) bool {
		if bgpPrefix.Address != nil && bgpPrefix.GetAddress() == expectedOTGBGPPrefix.Address &&
			bgpPrefix.PrefixLength != nil && bgpPrefix.GetPrefixLength() == expectedOTGBGPPrefix.PrefixLength {
			return true
		}
		return false
	})
```

Reply:
@DanG100 We have tried the following snippet with some modification, because we cann't do bgpPrefix.GetAddress() directly we have to do bgpPrefix.Val(t).GetAddress() other wise those functions are not directly available on that given struct.

`func verifyOTGBGP4Prefix(t *testing.T, otg *otg.OTG, config gosnappi.Config, expectedOTGBGPPrefix OTGBGPPrefix) {
	_, ok := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch(
		t,
		10*time.Second,
		func(bgpPrefix *otgtelemetry.QualifiedBgpPeer_UnicastIpv4Prefix) bool {
			return bgpPrefix.Val(t) != nil && bgpPrefix.Val(t).Address != nil && bgpPrefix.Val(t).GetAddress() == expectedOTGBGPPrefix.Address &&
				bgpPrefix.Val(t).PrefixLength != nil && bgpPrefix.Val(t).GetPrefixLength() == expectedOTGBGPPrefix.PrefixLength
		}).Await(t)

	if !ok {
		actPrefixes := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Get(t)
		for _, actPrefix := range actPrefixes {
			t.Logf("Peer: %v, Address: %v, Prefix Length: %v", expectedOTGBGPPrefix.PeerName, actPrefix.GetAddress(), actPrefix.GetPrefixLength())
		}
		t.Errorf("Given BGP IPv4 Prefix is not formed %v", expectedOTGBGPPrefix)
	}
}`

even though given prefix is present this condition is never got satisfied. we ensured from the snippet output as well.

`route_propagation_test.go:489: Checking for BGP Prefix 198.51.100.0/32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:292: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.0 32}
    route_propagation_test.go:489: Checking for BGP Prefix 198.51.100.1/32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:292: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.1 32}
    route_propagation_test.go:489: Checking for BGP Prefix 198.51.100.2/32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:292: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.2 32}
    route_propagation_test.go:489: Checking for BGP Prefix 198.51.100.3/32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:290: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:292: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.3 32}`

could you guys please give some pointer what we have done wrong, we also pushed this experimental function in the PR, so you can take a look there as well.

Reply:
Can you try a predicate like this:
```
return bgpPrefix.IsPresent() && bgpPrefix.Val(t).GetAddress() == expectedOTGBGPPrefix.Address &&
     bgpPrefix.Val(t).GetPrefixLength() == expectedOTGBGPPrefix.PrefixLength
```

Reply:
@DanG100 @liulk 

I tried the predicate like you suggested , but issue is still there,

```
func verifyOTGBGP4Prefix(t *testing.T, otg *otg.OTG, config gosnappi.Config, expectedOTGBGPPrefix OTGBGPPrefix) {
	_, ok := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch(
		t,
		10*time.Second,
		func(bgpPrefix *otgtelemetry.QualifiedBgpPeer_UnicastIpv4Prefix) bool {
			return bgpPrefix.IsPresent() && bgpPrefix.Val(t).GetAddress() == expectedOTGBGPPrefix.Address &&
				bgpPrefix.Val(t).GetPrefixLength() == expectedOTGBGPPrefix.PrefixLength
		}).Await(t)

	if !ok {
		actPrefixes := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Get(t)
		for _, actPrefix := range actPrefixes {
			t.Logf("Peer: %v, Address: %v, Prefix Length: %v", expectedOTGBGPPrefix.PeerName, actPrefix.GetAddress(), actPrefix.GetPrefixLength())
		}
		t.Errorf("Given BGP IPv4 Prefix is not formed %v", expectedOTGBGPPrefix)
	}
}
```

```
    route_propagation_test.go:471: Checking for BGP Prefix 198.51.100.0/32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:280: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.0 32}
    route_propagation_test.go:471: Checking for BGP Prefix 198.51.100.1/32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:280: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.1 32}
    route_propagation_test.go:471: Checking for BGP Prefix 198.51.100.2/32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:280: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.2 32}
    route_propagation_test.go:471: Checking for BGP Prefix 198.51.100.3/32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:278: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:280: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.3 32}
```

Please let us know, what we can tryout next, otherwise we are stuck at this point.

Reply:
@DanG100 could you please provide your feedback?

Reply:
What topology and environment are you using to run this test? I cannot reproduce because i'm running a different issue `PushConfig(t) on ATE{Dims:Dims{Name:ate Vendor:IXIA HardwareModel:IXIA_TG SoftwareVersion:IXIA_TG Ports:map[port1:Port{Name:eth2 Speed:SPEED_UNSPECIFIED CardModel: PMD:PMD_UNSPECIFIED} port2:Port{Name:eth1 Speed:SPEED_UNSPECIFIED CardModel: PMD:PMD_UNSPECIFIED}]}}: rpc error: code = Internal desc = Exception deserializing request`

Reply:
Can you try this?

```
	_, ok := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch(
		t,
		10*time.Second,
		func(bgpPrefix *otgtelemetry.QualifiedBgpPeer_UnicastIpv4Prefix) bool {
			if !bgpPrefix.isPresent() {
				return false
			}
			val := bgpPrefix.Val(t)
			addr, plen := val.GetAddress(), val.GetPrefixLength()
			t.Log("Any peer got: %s/%d", addr, plen)
			return addr == expectedOTGBGPPrefix.Address && plen == expectedOTGBGPPrefix.PrefixLength
		}).Await(t)
```

Reply:
We're using kne environment with [this](https://github.com/open-traffic-generator/featureprofiles/blob/6a66c7edeccd5d80809072129839fccf5cdb2a18/topologies/kne/arista_ixia.textproto) topology file and [this](https://github.com/open-traffic-generator/featureprofiles/blob/6a66c7edeccd5d80809072129839fccf5cdb2a18/topologies/atedut_2.testbed#L1) ondatra test bed
The test is [here](https://github.com/open-traffic-generator/featureprofiles/blob/c3bdd343ae06a7f009a2b03648ad402910d5802d/feature/bgp/addpath/otg_tests/route_propagation_test/route_propagation_test.go) and you can run it with this command 
`go test -v feature/bgp/addpath/otg_tests/route_propagation_test/*.go -kne-config ~/featureprofiles/topologies/kne/testbed.kne.yml -testbed ~/featureprofiles/topologies/atedut_2.testbed --deviation_interface_enabled --deviation_default_network_instance=default`

Reply:
The issue is that the OTG() is not sending a gNMI sync_response. 
For Watch and WatchAll on non-leaf paths, the predicate is only evaluated after a sync_response is sent (then evaluated on all subsequent notifications).

This can be verified by increasing the log verbosity using "--alsologtostderr -v=2"

Reply:
@liulk I have tried your suggestion with the code below
```func verifyOTGBGP4Prefix(t *testing.T, otg *otg.OTG, config gosnappi.Config, expectedOTGBGPPrefix OTGBGPPrefix) {
	lastValue, ok := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch(
		t,
		10*time.Second,
		func(bgpPrefix *otgtelemetry.QualifiedBgpPeer_UnicastIpv4Prefix) bool {
			if !bgpPrefix.IsPresent() {
				t.Log("Any peer not present")
				return false
			}
			val := bgpPrefix.Val(t)
			addr, plen := val.GetAddress(), val.GetPrefixLength()
			t.Logf("Any peer got: %s/%d", addr, plen)
			return addr == expectedOTGBGPPrefix.Address && plen == expectedOTGBGPPrefix.PrefixLength
		}).Await(t)

	if !ok {
		t.Logf("Last value was: %v", lastValue)
		actPrefixes := otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Get(t)
		for _, actPrefix := range actPrefixes {
			t.Logf("Peer: %v, Address: %v, Prefix Length: %v", expectedOTGBGPPrefix.PeerName, actPrefix.GetAddress(), actPrefix.GetPrefixLength())
		}
		t.Errorf("Given BGP IPv4 Prefix is not formed %v", expectedOTGBGPPrefix)
	}
}
```
but the predicate was never satisfied ... as you can see the last value returned by Await() was nil
```
    route_propagation_test.go:477: Checking for BGP Prefix 198.51.100.0/32
    route_propagation_test.go:268: Last value was: <nil>
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:273: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.0 32}
    route_propagation_test.go:477: Checking for BGP Prefix 198.51.100.1/32
    route_propagation_test.go:268: Last value was: <nil>
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:273: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.1 32}
    route_propagation_test.go:477: Checking for BGP Prefix 198.51.100.2/32
    route_propagation_test.go:268: Last value was: <nil>
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:273: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.2 32}
    route_propagation_test.go:477: Checking for BGP Prefix 198.51.100.3/32
    route_propagation_test.go:268: Last value was: <nil>
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.0, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.1, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.2, Prefix Length: 32
    route_propagation_test.go:271: Peer: port2.dev.BGP4.peer, Address: 198.51.100.3, Prefix Length: 32
    route_propagation_test.go:273: Given BGP IPv4 Prefix is not formed {port2.dev.BGP4.peer 198.51.100.3 32}

```

Reply:
@DanG100 looks like the predicate was never called with the Any watcher? @octpetre can you try running with the flag `-vmodule=genutil=1`?

I'm open to reverting to waitFor if we still can't figure out what's wrong with the verbose logging.

Reply:
the predicate is never called because the OTG doesn't send a gNMI sync_response. You may also need to --alsologtostderr flag. https://pkg.go.dev/github.com/golang/glog#section-documentation

Reply:
@octpetre sounds like an OTG bug to me. Can you file an OTG issue and reference it here? We can revert back to waitFor but let's add a comment about the desired approach and a reference to the issue blocking it.

```
// Use otg.Telemetry().BgpPeer(expectedOTGBGPPrefix.PeerName).UnicastIpv4PrefixAny().Watch()
// after this bug is fixed: https://...
```